common/src/types/v0/store/node.rs

```diff
@@ -66,33 +165,242 @@ impl NodeSpec {
+                                    state.cordonlabels,
+                                )));
+                            }
+                        } else {
+                            self.drain_state = Some(DrainState::Drained(state));
+                        }
+                    }
+                },
+                None => {
+                    // should not be possible
+                }
+            }
+        } else {
+            // return error
         }
     }
+
     /// Returns whether or not the node is cordoned.
     pub fn cordoned(&self) -> bool {
-        !self.cordon_labels.is_empty()
+        match self.drain_state.clone() {
+            Some(ds) => match ds {
+                DrainState::Cordoned(_) => true,
+                DrainState::Draining(_) => true,
+                DrainState::Drained(_) => true,
+            },
+            None => false,
+        }
     }
-    /// Returns the cordon labels
-    pub fn cordon_labels(&self) -> Vec<String> {
-        self.cordon_labels.clone()
+
+    pub fn has_cordon_label(&self, label: String) -> bool {
+        if self.has_cordon_only_label(label.clone()) || self.has_drain_label(label) {
+            return true;
+        }
+        false
+    }
+
+    pub fn has_cordon_only_label(&self, label: String) -> bool {
+        match self.drain_state.clone() {
+            Some(ds) => match ds {
+                DrainState::Draining(state) => state.cordonlabels.iter().any(|i| i == &label),
+                DrainState::Drained(state) => state.cordonlabels.iter().any(|i| i == &label),
+                DrainState::Cordoned(state) => state.cordonlabels.iter().any(|i| i == &label),
+            },
+            None => false,
+        }
+    }
+
+    pub fn has_drain_label(&self, label: String) -> bool {
+        match self.drain_state.clone() {
+            Some(ds) => match ds {
+                DrainState::Draining(state) => state.drainlabels.iter().any(|i| i == &label),
+                DrainState::Drained(state) => state.drainlabels.iter().any(|i| i == &label),
+                DrainState::Cordoned(_) => false,
+            },
+            None => false,
+        }
+    }
+
+    pub fn is_draining(&self) -> bool {
+        match self.drain_state.clone() {
+            Some(ds) => match ds {
+                DrainState::Cordoned(_) => false,
+                DrainState::Draining(_) => true,
+                DrainState::Drained(_) => false,
+            },
+            None => false,
+        }
     }
 }
 
 impl From<NodeSpec> for models::NodeSpec {
     fn from(src: NodeSpec) -> Self {
-        Self::new(src.endpoint, src.id, src.cordon_labels)
+        if src.drain_state.is_none() {
+            Self::new(src.endpoint, src.id)
+        } else {
+            let openapi_ds: Option<models::DrainState> = src.drain_state.map(|ds| ds.into());
+            Self::new_all(src.endpoint, src.id, openapi_ds)
```

Comment:
might be able to just use `new_all(..., src.drain_state.into_opt())`

Reply:
fixed